httpx/auth.py

```diff
@@ -36,3 +40,132 @@ def build_auth_header(self) -> str:
         userpass = b":".join((username, password))
         token = b64encode(userpass).decode().strip()
         return f"Basic {token}"
+
+
+class HTTPDigestAuth:
+
+    ALGORITHM_TO_HASH_FUNCTION = {
+        "MD5": hashlib.md5,
+        "SHA": hashlib.sha1,
+        "SHA-256": hashlib.sha256,
+        "SHA-512": hashlib.sha512,
+    }
+
+    def __init__(
+        self, username: typing.Union[str, bytes], password: typing.Union[str, bytes]
+    ) -> None:
+        self.username = username
+        self.password = password
+
+    def __call__(self, request: AsyncRequest, response) -> AsyncRequest:
+        auth_header = self.build_auth_header(request, response)
+        if auth_header is not None:
+            # TODO: raise instead of returning None?
+            request.headers["Authorization"] = auth_header
+        return request
+
+    def build_auth_header(
+        self, request: AsyncRequest, response: AsyncResponse
+    ) -> typing.Optional[str]:
+        # Retrieve challenge from response header
+        header = response.headers.get("www-authenticate")
+        assert header.lower().startswith("digest")
+        challenge = self._parse_header(header)
+        algorithm = challenge.get("algorithm", "MD5")
+
+        realm = _safe_encode(challenge["realm"])
+        nonce = _safe_encode(challenge["nonce"])
+        qop = _safe_encode(challenge["qop"]) if "qop" in challenge else None
+        opaque = _safe_encode(challenge["opaque"]) if "opaque" in challenge else None
+        username = _safe_encode(self.username)
+        password = _safe_encode(self.password)
+
+        # Assemble parts depending on hash algorithms
+        hash_func = self.ALGORITHM_TO_HASH_FUNCTION[
+            algorithm.replace("-SESS", "").upper()
+        ]
+
+        def digest(data: bytes) -> bytes:
+            return hash_func(data).hexdigest().encode("utf-8")
+
+        def keyed_digest(secret, data) -> bytes:
+            return digest(b":".join((secret, data)))
+
+        A1 = b":".join((username, realm, password))
+        HA1 = digest(A1)
+
+        path = self._get_parsed_path(request.url)
+        A2 = b":".join((request.method.encode("utf-8"), path))
+        # TODO: implement auth-int
```

Comment:
I believe [Requests does not currently implement `auth-int`](https://github.com/kennethreitz/requests/blob/master/requests/auth.py#L210). Do we want to?

Reply:
Treat it as out-of-scope for now.